File: backend/app/api/medicines.py
# backend/app/api/medicines.py
"""
API endpoints for fetching medicines from drugs_database collection.
"""

from fastapi import APIRouter, Query
from app.db.mongo_db import mongodb_service
from typing import Optional
import logging

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/medicines", tags=["medicines"])


@router.get("")
async def get_medicines(
    limit: int = Query(50, ge=1, le=200),
    skip: int = Query(0, ge=0),
    search: Optional[str] = Query(None)
):
    """
    Fetch medicines from the drugs_database collection.
    
    Args:
        limit: Maximum number of medicines to return (default 50)
        skip: Number of medicines to skip for pagination
        search: Optional search query for drug_name or generic_name
    """
    try:
        db = mongodb_service.db
        if db is None:
            return {
                "success": False,
                "error": "Database not connected",
                "data": [],
                "total": 0
            }
        
        collection = db["drugs_database"]
        
        # Build query
        query = {}
        if search:
            query = {
                "$or": [
                    {"drug_name": {"$regex": search, "$options": "i"}},
                    {"generic_name": {"$regex": search, "$options": "i"}}
                ]
            }
        
        # Get total count
        total = await collection.count_documents(query)
        
        # Fetch medicines
        cursor = collection.find(query).skip(skip).limit(limit)
        medicines = []
        
        async for doc in cursor:
            medicines.append({
                "id": str(doc.get("_id")),
                "drug_name": doc.get("drug_name", "Unknown"),
                "generic_name": doc.get("generic_name", "Unknown"),
                "known_side_effects": doc.get("known_side_effects", []),
                "common_dosages": doc.get("common_dosages", []),
                "approved_countries": doc.get("approved_countries", [])
            })
        
        logger.info(f"Fetched {len(medicines)} medicines from drugs_database")
        
        return {
            "success": True,
            "data": medicines,
            "total": total,
            "limit": limit,
            "skip": skip
        }
        
    except Exception as e:
        logger.error(f"Error fetching medicines: {e}")
        return {
            "success": False,
            "error": str(e),
            "data": [],
            "total": 0
        }


@router.get("/count")
async def get_medicines_count():
    """Get total count of medicines in the database."""
    try:
        db = mongodb_service.db
        if db is None:
            return {"success": False, "count": 0, "error": "Database not connected"}
        
        collection = db["drugs_database"]
        count = await collection.count_documents({})
        return {"success": True, "count": count}
    except Exception as e:
        logger.error(f"Error counting medicines: {e}")
        return {"success": False, "count": 0, "error": str(e)}

File: Frontend/src/api/medicines.js
/**
 * Medicines API module
 * Connects to backend medicines endpoints
 */

const API_BASE = 'http://localhost:8000';

/**
 * Get medicines from the database
 * @param {Object} options - Query options
 * @param {number} options.limit - Maximum number of medicines (default 50)
 * @param {number} options.skip - Number to skip for pagination (default 0)
 * @param {string} options.search - Optional search query
 * @returns {Promise<{success: boolean, data: Array, total: number, limit: number, skip: number}>}
 */
export const getMedicines = async ({ limit = 50, skip = 0, search = '' } = {}) => {
    const params = new URLSearchParams({
        limit: limit.toString(),
        skip: skip.toString()
    });

    if (search) {
        params.append('search', search);
    }

    const response = await fetch(`${API_BASE}/api/v1/medicines?${params}`);

    if (!response.ok) {
        throw new Error(`Failed to fetch medicines: ${response.statusText}`);
    }

    return response.json();
};

/**
 * Get total count of medicines
 * @returns {Promise<{success: boolean, count: number}>}
 */
export const getMedicinesCount = async () => {
    const response = await fetch(`${API_BASE}/api/v1/medicines/count`);

    if (!response.ok) {
        throw new Error(`Failed to fetch medicines count: ${response.statusText}`);
    }

    return response.json();
};
